### DISCUSSION TITLE
How to reduce latency between tasks?

### DISCUSSION THREAD

**Discussion by h2cone** [apache/airflow#20337] [Q&A]
Gantt chart:
![1639384474708](https://user-images.githubusercontent.com/24283234/146346762-ae387c02-d6f9-4027-abae-df735eb5d0c7.jpg)

We only use [SimpleHttpOperator](https://github.com/apache/airflow/blob/main/airflow/providers/http/operators/http.py) or [ExtendedHttpOperator](https://gist.github.com/nndd91/56c2337b89e2b8705a90da088020b609), including on_success_callback and on_failure_callback, and the number of retries is 0.

Airflow version: 2.1.4
Configuration file: [docker-compose.yaml](https://airflow.apache.org/docs/apache-airflow/2.1.4/docker-compose.yaml)

**Comment by potiuk** [comment]
Are you using Sequential Executor and SQLite maybe ? If so - don't.  
Switch to a "regular db" (Postgres is recommended) and LocalExecutor.
 
If you use postgres already you need to take a look at https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html?highlight=scheduler#fine-tuning-your-scheduler-performance

**Reply by h2cone** [reply]
> Are you using Sequential Executor and SQLite maybe ? If so - don't. Switch to a "regular db" (Postgres is recommended) and LocalExecutor.
> 
> If you use postgres already you need to take a look at https://airflow.apache.org/docs/apache-airflow/stable/concepts/scheduler.html?highlight=scheduler#fine-tuning-your-scheduler-performance

Thanks for replying. I am using LocalExecutor instead of CeleryExecutor and the database is Postgres.

I read the article and tried to configure the options, but I found that the latency between tasks is still uncontrollable, mostly between 1 and 2 seconds, with the longest being around 5 seconds.

![1640230761019](https://user-images.githubusercontent.com/24283234/147184549-82c6a513-1bd9-4f20-88af-b0e619242104.jpg)

What happened between tasks? How to find out the cause of the delay between tasks?

**Reply by potiuk** [reply]
No idea why you have it. In all my tests with Postgres DB Airlfow run super fast with pretty much 0 delay between tasks. You need to see (following the article) where your delays are - maybe you have super-slow filesystem or remote database that takes 0.5 second to run query on. Or maybe your scheduler gets continuously restarted because you have low memory. Or maybe you have other processes running on the same machine that competing for the same resources. Or maybe your network is super slow. Or disk. There might be many reasons. 

Docker Compose is not the best deployment environment. As we warned our users here:  https://airflow.apache.org/docs/apache-airflow/stable/start/docker.html#production-readiness:

> short of becoming a Docker Compose expert, it’s highly unlikely you will get robust deployment with it.

 if you choose docker-compose environment, there is nothing we can help you with to fine-tune your environment - you need to become docker-compose expert if you want to "productionize" it. So you are, I am afraid, pretty much on your own.

You can very esily check it and run "standalone" version of airlfow on a decent machine and see that it runs with pretty much 0 delays https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#standalone . Then you will see that it's your deployment issue, not airflow. And it's pretty much up to you to investigate and fix the issue - you need to follow the article and see which resources are blocking you.

**Reply by AlwaysLearningMore** [reply]
@potiuk  When you say "with pretty much 0 delays".  How low are you able to achieve?  Do you have an image of a gantt chart with very small inter task latency?  I am considering deploying Airflow in my org. but want to understand what the limitations are on latency.

**Reply by potiuk** [reply]
You can watch this talk from Ash where he explains not only how Scheduler achieves that but also explains some benchmarks https://www.youtube.com/watch?v=DYC4-xElccE  also this astronomer blog has nice writeup and some benchmarks https://www.astronomer.io/blog/airflow-2-scheduler/

I think you can also find more by googling it. We don't have "official" benchmarks, but I know people published the, but if I were to find some better, I would have to google it the same way, so better if you do it.
